[VideoPlayer] Double the initial refresh rate for interlaced streams

CrystalP opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

Switch resolution/refresh rate only once when starting the playback of interlaced video streams, instead of twice with the resulting black frames and audio drop out / pitch change.

For example, this stream has the top field first flag and is detected as interlaced by ffmpeg analysis:
```
Stream #0:0(eng): Video: mpeg2video (Main), yuv420p(tv, top first), 720x576 [SAR 16:15 DAR 4:3], 25 fps, 25 tbr, 1k tbn (default) (forced)
```

Current behavior:
Initial switch to a resolution with 25Hz refresh rate, followed a few seconds later by a switch to 50Hz.
Desired behavior:
Initial switch to 50Hz. No additional resolution changes.

To accomplish that, propagate the field order information from the ffmpeg analysis results to the hints used when a stream is opened, so that VideoPlayer can double the initial refresh rate for interlaced streams.


.ts stream handling in demuxer is different to enable fast switching for PVR addons and has a complicated history (see #5487 and #3590). The consequence is that the ffmpeg analysis results are lost and this PR is unable to propagate the interlaced flags and to remove the additional resolution change.
#19983 added an exception for hevc which doesn't seem to have disturbed PVR (likely because they only stream H264?)

Addressing that / revisiting .ts stream opening is beyond the scope of this PR and close to a release is not a good time for that kind of risky changes anyway.

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

Unnecessary resolution switch for streams detected as interlaced.
Noticed personally and was brought up in forum thread https://forum.kodi.tv/showthread.php?tid=375049

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Used the sample provided by user in the forum thread.
Regression testing with other samples that are detected as unknown or progressive field order.

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

Remove extra resolution change when starting playback of interlaced streams.
Except .ts

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

CrystalP committed:
[VideoPlayer] Use probed interlaced flags to double initial refresh rate

When starting the playback of an interlaced stream, instead of switching to the probed refresh rate,
go directly to 2x the probed refresh rate as videoplayer would otherwise switch to 2x refresh rate
a few seconds later anyway, resulting in some black frames while the screen adjusts and audio drop out or a pitch change.

Does not work for ts streams because of special opening by demuxer, which loses most of the probed data.
Addressing that is beyond the scope of this change.

smp79 commented:
It didn't fix the double resolution change for me.

[debug log](https://hastebin.com/share/pojitedazu.yaml)

CrystalP commented:
> It didn't fix the double resolution change for me.

Thanks for the feedback and the log, I also noticed an issue on a .ts file and there is special handling of that type by Kodi's demuxer which seems to lose the field order information. I'll take a closer look, but it's likely outside the scope of this change and there likely were reasons it was done that way.

CrystalP commented:
I looked closer at the .ts behavior and it's the consequence of a change for fast channel switching of PVR addons. There is some complicated history available in #5487 and #3590 and I'm not touching that with a 10 foot pole close to release.

Squashed the commits, modified the commit messages a bit, I think it's ready for merge.

CrystalP commented:
Jenkins build this please

reardonia commented:
@CrystalP sorry i know this is ancient history but I'm trying to understand why you did the fps doubling here?  currently ffmpeg reports out 50fps for 576i25 video, which causes Kodi (this PR) to again double to 100fps.  is that all intentional?

CrystalP commented:
Just tried a few samples on current master/ffmpeg9 and no change. ffmpeg still reports the frame rate as 25 or 29.97, so the single doubling of the PR works as expected.
The special case of .ts streams remains, with its incomplete second ffmpeg probe that doesn't report interlacing so no doubling.

CrystalP commented:
could it be a variation of the slightly off framerate at the mkv container level?

reardonia commented:
@CrystalP I am using what I think is well-known interlaced test clip, "burosch"  MPEG2_576i_25-interlace_burosch1.mkv

excerpt:

```
14607:2026-08-12 18:23:47.962 T:6158     info <general>: VideoPlayer::OpenFile: smb://mediasrv/Download/test-videos/MPEG2_576i_25-interlace_burosch1.mkv
14637:2026-08-12 18:23:47.989 T:6592     info <general>: ffmpeg[0x5f53d24f8790]:   Stream #0:0(ger): Video: mpeg2video (Main), yuv420p(tv, bottom first), 720x576 [SAR 64:45 DAR 16:9], 8000 kb/s, 50 fps, 50 tbr, 1k tbn (default)
14655:2026-08-12 18:23:47.989 T:6592    debug <general>: CDVDDemuxFFmpeg::AddStream ID: 0
14658:2026-08-12 18:23:47.989 T:6592     info <general>: Opening stream: 0 source: 256
14659:2026-08-12 18:23:47.989 T:6592     info <general>: [WHITELIST] Searching the whitelist for: width: 720, height: 576, fps: 100.000, 3D: false
14675:2026-08-12 18:23:47.990 T:6592     info <general>: Display resolution ADJUST : 3840x1600 @ 74.977379 Hz (16) (weight: 0.000)
14687:2026-08-12 18:23:47.999 T:6158    debug <general>: CDVDDemuxFFmpeg::AddStream ID: 0
14694:2026-08-12 18:23:48.126 T:6592     info <general>: Creating video codec: mpeg2video
14697:2026-08-12 18:23:48.127 T:6592    debug <general>: CVideoPlayerVideo::OpenStream - open stream with codec: mpeg2video
14705:2026-08-12 18:23:48.130 T:6158     info <general>: [WHITELIST] Searching the whitelist for: width: 720, height: 576, fps: 100.000, 3D: false
14721:2026-08-12 18:23:48.130 T:6158    debug <general>: [WHITELIST] No match for an exact resolution with an exact refresh rate
14722:2026-08-12 18:23:48.130 T:6158    debug <general>: [WHITELIST] Searching for an exact resolution with double the refresh rate
14723:2026-08-12 18:23:48.131 T:6158    debug <general>: [WHITELIST] No match for an exact resolution with double the refresh rate
14724:2026-08-12 18:23:48.131 T:6158    debug <general>: [WHITELIST] Searching for an exact resolution with a 3:2 pulldown refresh rate
14725:2026-08-12 18:23:48.131 T:6158    debug <general>: [WHITELIST] No match for a resolution with a 3:2 pulldown refresh rate
14726:2026-08-12 18:23:48.131 T:6158    debug <general>: [WHITELIST] Searching for a desktop resolution with an exact refresh rate
14727:2026-08-12 18:23:48.131 T:6158    debug <general>: [WHITELIST] No match for a desktop resolution with an exact refresh rate
14728:2026-08-12 18:23:48.131 T:6158    debug <general>: [WHITELIST] Searching for a desktop resolution with double the refresh rate
14731:2026-08-12 18:23:48.132 T:6158    debug <general>: [WHITELIST] No match for a desktop resolution with double the refresh rate
14732:2026-08-12 18:23:48.132 T:6158    debug <general>: [WHITELIST] Searching for a desktop resolution with a 3:2 pulldown refresh rate
14733:2026-08-12 18:23:48.132 T:6158    debug <general>: [WHITELIST] No match for a desktop resolution with a 3:2 pulldown refresh rate
14734:2026-08-12 18:23:48.132 T:6158    debug <general>: [WHITELIST] No resolution matched
14735:2026-08-12 18:23:48.132 T:6158     info <general>: Display resolution ADJUST : 3840x1600 @ 74.977379 Hz (16) (weight: 0.000)
14764:2026-08-12 18:23:48.156 T:6596    debug <general>: CRenderManager::Configure - change configuration. 720x576. display: 720x405. framerate: 50.00.
14831:2026-08-12 18:23:48.186 T:6158     info <general>: [WHITELIST] Searching the whitelist for: width: 720, height: 576, fps: 50.000, 3D: false
14840:2026-08-12 18:23:48.187 T:6158    debug <general>: [WHITELIST] Matched a desktop resolution with an exact refresh rate 3840x2160 @ 50.000000 Hz (20)
14841:2026-08-12 18:23:48.187 T:6158     info <general>: Display resolution ADJUST : 3840x2160 @ 50.000000 Hz (20) (weight: 0.000)
14920:2026-08-12 18:24:10.595 T:6596    debug <general>: CalcFrameRate counted 1000 frames without being able to calculate the framerate, giving up
```

ffmpeg says 50fps, and then Kodi (master) searches for 100fps based on this PR (I think)


```
$ mediainfo MPEG2_576i_25-interlace_burosch1.mkv
General
Unique ID                                : 195365263808220068521172837861940898953 (0x92F9FB2DB4335312BA8245D9CDBDAC89)
Complete name                            : MPEG2_576i_25-interlace_burosch1.mkv
Format                                   : Matroska
Format version                           : Version 4
File size                                : 18.4 MiB
Duration                                 : 30 s 0 ms
Overall bit rate mode                    : Constant
Overall bit rate                         : 5 132 kb/s
Frame rate                               : 25.000 FPS
Encoded date                             : 2017-01-17 16:40:39 UTC
Writing application                      : mkvmerge 9.7.1 ('Pandemonium') 32bit
Writing library                          : libebml v1.3.4 + libmatroska v1.4.5

Video
ID                                       : 1
Format                                   : MPEG Video
Format version                           : Version 2
Format profile                           : Main@Main
Format settings                          : BVOP
Format settings, BVOP                    : Yes
Format settings, Matrix                  : Default
Format settings, GOP                     : M=3, N=12
Format settings, picture structure       : Frame
Codec ID                                 : V_MPEG2
Codec ID/Info                            : MPEG 1 or 2 Video
Duration                                 : 30 s 0 ms
Bit rate mode                            : Constant
Bit rate                                 : 5 126 kb/s
Maximum bit rate                         : 8 000 kb/s
Width                                    : 720 pixels
Height                                   : 576 pixels
Display aspect ratio                     : 16:9
Frame rate mode                          : Constant
Frame rate                               : 25.000 FPS
Standard                                 : PAL
Color space                              : YUV
Chroma subsampling                       : 4:2:0
Bit depth                                : 8 bits
Scan type                                : Interlaced
Scan order                               : Bottom Field First
Compression mode                         : Lossy
Bits/(Pixel*Frame)                       : 0.494
Time code of first frame                 : 00:00:00:00
Time code source                         : Group of pictures header
GOP, Open/Closed                         : Closed
Stream size                              : 18.3 MiB (100%)
Language                                 : German
Default                                  : Yes
Forced                                   : No
```

CrystalP commented:
I wasn't aware of that clip, would you send me a copy in slack.
